### PR TITLE
BUG Handle Categorical columns with non-string types

### LINKED ISSUE #18: BUG DataFrameETL fails with int-like Categoricals
The following fails under v0.1.5 (the most recent release):

```python
    raw = pd.concat([
        pd.Series([1.0, np.NaN, 3.0], dtype='float', name='fruits'),
        pd.Series([500, 1000, 1000], dtype='category', name='intcat'),
    ], axis=1)
    expander = DataFrameETL(cols_to_expand='auto', dummy_na=True)
    tfm = expander.fit_transform(raw)
```

The error is "ValueError: fill value must be in categories".

It looks like this is due to `DataFrameETL._flag_numeric` incorrectly marking a `pd.Categorical` as "numeric" when every level happens to be an integer.

### PR DESCRIPTION
**Pull request by stephen-hoover**
We previously could not expand `pd.Categorical` columns which had non-string levels. A column with all integer levels would be marked as a numeric column, causing us to try to fill missing values with our sentinel on the column itself. This errored because the sentinel wasn't part of the existing levels. Mixed types would fail because the expansion code did sorting operations, and numbers have no defined order relative to strings.

This commit refactors the expansion code to cast columns to `pd.Categorical` types and use `pd.get_dummies`, still on a column-wise basis. Previous behavior is preserved, but now the `DataFrameETL` should handle all valid `pd.Categorical` columns in the input. This also allows some simplification in the code. Since we're handling columns as categoricals, we can use the same sentinel for all columns.

Closes #18 .

### CODE REVIEW

**Comment by beckermr** [comment]
We need to make sure to throw a really big DataFrame at this. `pd.get_dummies` scares me. @jmemich, didn't you have trouble with it before?

**Comment by beckermr** [comment]
Also, why do we need the sentinels if we use a categorical type in pandas? I think we can just keep it as np.NaN everywhere, right? The sentinels were something I invented to be able to use sklearn one-hot-encoding since it does not handle NaNs. We do not need this if we use `pd.get_dummies`.

**Comment by beckermr** [comment]
One more idea. Can we just cast all categorical columns to object types internally when we process them and leave the rest of the code as is?

**Comment by stephen-hoover** [comment]
IIRC, the resource problem with `get_dummies` came when we used it on the entire DataFrame at once.

I kept the sentinel because the current behavior of the `DataFrameETL` is to completely ignore new categories found at transform time. The simplest way to make sure that we have the same levels when transforming as for fitting is to set the levels on the columns to transform to be the same as the levels seen when fitting. In this case, `pandas` treats new categories as missing data. Using the sentinel avoids this confusion. There would be ways to handle this without a sentinel, but I think it would involve extra mucking with the categories and removing extra columns after expansion. I think the code would end up being more complicated that way.

My first solution involved casting columns to object types, but you're right, that still runs into errors when trying to sort.

**Comment by jmemich** [comment]
FWIW, I was looking the `pandas` source code and calling `pd.get_dummies` one column at a time avoids [this copy](https://github.com/pandas-dev/pandas/blob/master/pandas/core/reshape/reshape.py#L829-L832), which may have rendered one of my earlier implementations unworkable.

**Comment by beckermr** [comment]
Thanks @stephen-hoover and @jmemich! I am going to throw my standard test problem at this before we merge to make sure we don't have any hidden performance degradations. I am more convinced this is a good solution now.

**Comment by jmemich** [comment]
> my standard test problem

@beckermr if you have a script that's in decent enough shape I think it would make sense to add it to a `benchmarks/` directory. This kind of information is a good thing to tie to the codebase, *in* the codebase.

File: civismlext/test/test_preprocessing.py [hunk 1]
```diff
@@ -677,6 +635,22 @@ def test_pickle(data_raw, dataframe_expected):
     expander = pickle.load(buff)
 
     arr = expander.transform(data_raw)
-    expected_array = np.asarray(dataframe_expected)
     assert arr.shape == expected_array.shape
     assert_almost_equal(arr, expected_array)
+
+
+def test_categorical_looks_like_int():
```
Comment (stephen-hoover) [inline civismlext/test/test_preprocessing.py]: This test fails before the changes in this PR.

**Comment by beckermr** [comment]
Good point @jmemich. I can try and throw something together.

File: civismlext/preprocessing.py [hunk 2]
```diff
@@ -176,33 +86,14 @@ def _flag_nulls(self, X, cols_to_drop):
                                  '"raise", or "warn"]')
         return cols_to_drop + null_cols
 
-    def _flag_numeric(self, levels):
```
Comment (elsander) [inline civismlext/preprocessing.py]: So excited to see so many LOC go away!

File: civismlext/test/test_preprocessing.py (hunk 1 above)
Comment (elsander) [reply civismlext/test/test_preprocessing.py]: Should we include a mixed-type categorical here as well?

File: civismlext/test/test_preprocessing.py [hunk 3]
```diff
@@ -677,6 +635,22 @@ def test_pickle(data_raw, dataframe_expected):
...
+    # Verify that the right thing happens if the fit DataFrame has a
+    # categorical with
```
Comment (elsander) [inline civismlext/test/test_preprocessing.py]: It looks like this comment got cut off.

File: civismlext/test/test_preprocessing.py (hunk 1 above)
Comment (beckermr) [reply civismlext/test/test_preprocessing.py]: yes!

File: civismlext/test/test_preprocessing.py (hunk 3 above)
Comment (stephen-hoover) [reply civismlext/test/test_preprocessing.py]: Fixed.

File: civismlext/test/test_preprocessing.py (hunk 1 above)
Comment (stephen-hoover) [reply civismlext/test/test_preprocessing.py]: You're right, we should. Added!

**Comment by beckermr** [comment]
Alright. I did some testing and am super stoked. This appears to be 3x faster with no memory penalty. LGTM!

**Comment by stephen-hoover** [comment]
I did some testing of my own. Transforming 4 columns of an input table with ~350 columns, (~170 total categories present at the larger two sizes) I find memory use of:

| N rows | v0.1.5 | This PR |
|---------|-------|---------|
| 10,000 | 255 MiB | 255 MiB|
|100,000 | 850 MiB | 850 MiB |
|1,000,000 | 7500 MiB | 7200 MIB |

I'm measuring memory use in Civis Platform scripts. Note that these numbers vary a bit from run to run. I find no evidence for increased memory usage, consistent with Matt's tests.

When I timed the 1M row expansion, the transform step took ~11.9 s on average (over 100 tries) with v0.1.5, and ~6.7 s on average with the code from this branch.

**Pull request merged**